src/drivers/ammo-driver.js

```diff
@@ -1,10 +1,174 @@
+window.AmmoModule = window.Ammo;
+window.Ammo = null;
 
-function AmmoDriver () {
-  throw new Error('[AmmoDriver] Driver not implemented.');
+const EPS = 10e-6;
+
+function AmmoDriver() {
+  this.collisionConfiguration = null;
+  this.dispatcher = null;
+  this.broadphase = null;
+  this.solver = null;
+  this.physicsWorld = null;
+  this.debugDrawer = null;
+
+  this.els = {};
+  this.eventListeners = [];
+  this.collisions = {};
+  this.currentCollisions = [];
 }
 
 AmmoDriver.prototype = new Driver();
 AmmoDriver.prototype.constructor = AmmoDriver;
 
 module.exports = AmmoDriver;
+
+/* @param {object} worldConfig */
+AmmoDriver.prototype.init = function(worldConfig) {
+  return new Promise(resolve => {
+    AmmoModule().then(result => {
+      Ammo = result;
+      this.epsilon = worldConfig.epsilon || EPS;
+      this.debugDrawMode = worldConfig.debugDrawMode || THREE.AmmoDebugConstants.NoDebug;
+      this.maxSubSteps = worldConfig.maxSubSteps || 4;
+      this.fixedTimeStep = worldConfig.fixedTimeStep || 1 / 60;
+      this.collisionConfiguration = new Ammo.btDefaultCollisionConfiguration();
+      this.dispatcher = new Ammo.btCollisionDispatcher(this.collisionConfiguration);
+      this.broadphase = new Ammo.btDbvtBroadphase();
+      this.solver = new Ammo.btSequentialImpulseConstraintSolver();
+      this.physicsWorld = new Ammo.btDiscreteDynamicsWorld(
+        this.dispatcher,
+        this.broadphase,
+        this.solver,
+        this.collisionConfiguration
+      );
+      this.physicsWorld.setGravity(
+        new Ammo.btVector3(0, worldConfig.hasOwnProperty("gravity") ? worldConfig.gravity : -9.8, 0)
+      );
+      this.physicsWorld.getSolverInfo().set_m_numIterations(worldConfig.solverIterations);
+      resolve();
+    });
+  });
+};
+
+/* @param {Ammo.btCollisionObject} body */
+AmmoDriver.prototype.addBody = function(body, group, mask) {
+  this.physicsWorld.addRigidBody(body, group, mask);
+  this.els[Ammo.getPointer(body)] = body.el;
+};
+
+/* @param {Ammo.btCollisionObject} body */
+AmmoDriver.prototype.removeBody = function(body) {
+  this.physicsWorld.removeRigidBody(body);
+  this.removeEventListener(body);
+  delete this.els[Ammo.getPointer(body)];
+};
+
+AmmoDriver.prototype.updateBody = function(body) {
+  if (this.els.hasOwnProperty(Ammo.getPointer(body))) {
+    this.physicsWorld.updateSingleAabb(body);
+  }
+};
+
+/* @param {number} deltaTime */
+AmmoDriver.prototype.step = function(deltaTime) {
+  this.physicsWorld.stepSimulation(deltaTime, this.maxSubSteps, this.fixedTimeStep);
+
+  const numManifolds = this.dispatcher.getNumManifolds();
+  for (let i = 0; i < numManifolds; i++) {
+    const persistentManifold = this.dispatcher.getManifoldByIndexInternal(i);
+    const numContacts = persistentManifold.getNumContacts();
+    const body0ptr = Ammo.getPointer(persistentManifold.getBody0());
+    const body1ptr = Ammo.getPointer(persistentManifold.getBody1());
+    const key = body0ptr + "_" + body1ptr;
```

Comment:
Does Ammo guarantee that manifolds representing a collision of bodies A and B will consistently assign them to be the first and second body in the same order every time? If not, this would be broken.

Reply:
I have no evidence other than that from running this code and printing out those pointers, they always seem to be in the same order.

Reply:
Hmm, maybe I will read the Bullet code and try to figure out whether it's always going to be true or whether we will get tricked.